src/Processors/Transforms/Streaming/JoinTransformWithAlignment.cpp

```diff
@@ -24,113 +31,78 @@ JoinTransformWithAlignment::JoinTransformWithAlignment(
 
     /// We know the finalized left header, output header etc, post init HashJoin
     join->postInit(left_input_header, output_header, join_max_cached_bytes);
-    assert(join->requireWatermarkAlignedStreams());
 
     auto left_join_stream_desc = join->leftJoinStreamDescription();
-    latency_threshold = left_join_stream_desc->latency_threshold;
-    assert(latency_threshold != 0);
+    auto right_join_stream_desc = join->rightJoinStreamDescription();
 
-    /// When left stream or right stream is in quiesce, we still need to progress the join probably,
-    /// Otherwise we may stuck forever.
-    quiesce_threshold_ms = left_join_stream_desc->quiesce_threshold_ms;
-    if (quiesce_threshold_ms <= 0)
-        quiesce_threshold_ms = std::max<Int64>(latency_threshold * 2, 500);
+    left_input.watermark_column_position = left_join_stream_desc->alignmentKeyColumnPosition();
+    left_input.watermark_column_type = left_join_stream_desc->alignmentKeyColumnType();
 
-    auto right_join_stream_desc = join->rightJoinStreamDescription();
+    right_input.watermark_column_position = right_join_stream_desc->alignmentKeyColumnPosition();
+    right_input.watermark_column_type = right_join_stream_desc->alignmentKeyColumnType();
 
-    left_watermark_column_position = left_join_stream_desc->lagBehindColumnPosition();
-    left_watermark_column_type = left_join_stream_desc->lagBehindColumnType();
+    left_input.is_changelog_input = isChangelogDataStream(left_join_stream_desc->data_stream_semantic);
+    right_input.is_changelog_input = isChangelogDataStream(right_join_stream_desc->data_stream_semantic);
 
-    right_watermark_column_position = right_join_stream_desc->lagBehindColumnPosition();
-    right_watermark_column_type = right_join_stream_desc->lagBehindColumnType();
+    /// Some join combinations have builtin buffer to do aligning, so don't build an extra external buffer to align:
+    /// 1) For asof join, the right side has builtin buffer to keep multiple versions
+    /// 2) For range join, the left and right side both have builtin buffer to keep multiple ranges
+    left_input.need_aligned_buffer = !join->leftHasBuiltInAlignedBuffer();
```

Comment:
Here we like to say, if we like to buffer data in transform or just go ahead to push the data to hash join object.

We will need revise a few things
1. The interface names. 
2. The variable names

left_input.need_aligned_buffer => left_input.need_buffer_data_to_align
leftHasBuiltInAlignedBuffer => leftStreamRequiresBufferingDataToAlign

After we refactor out different join mode, these interface can go away probably.